cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,234 @@
+    Callable,
+    Tuple,
+    TypeVar,
+    Hashable,
+    Optional,
+    cast,
+    Type,
+    overload,
+)
+import dataclasses
+from enum import Enum
+
+from cirq.circuits.circuit import CIRCUIT_TYPE
+from cirq.circuits import circuit, frozen_circuit
+
+
+class LogLevel(Enum):
+    """Different logging resolution options for `TransformerStatsLogger`."""
+
+    INFO = 1
+    VERBOSE = 2
+    WARNING = 3
+
+
+class TransformerStatsLoggerBase:
+    """Abstract Base Class for transformer logging infrastructure.
+
+    The logger implementation should be stateful, s.t.:
+        - Each call to `register_initial` registers a new transformer stage and initial circuit.
+        - Each subsequent call to `log` should store additional logs corresponding to the stage.
+        - Each call to `register_final` should register the end of the currently active stage.
+
+    The logger assumes that
+        - Transformers are run sequentially.
+        - Nested transformers are allowed, in which case the behavior would be similar to a
+          doing a depth first search on the graph of transformers -- i.e. the top level transformer
+          would end (i.e. receive a `register_final` call) once all nested transformers (i.e. all
+          `register_initial` calls received while the top level transformer was active) have
+          finished (i.e. corresponding `register_final` calls have also been received).
+        - This behavior can be simulated by maintaining a stack of currently active stages and
+          adding data from `log` calls to the stage at the top of the stack.
+
+    The `LogLevel`s can be used to control the input processing and output resolution of the logs.
+    """
+
+    @abc.abstractmethod
+    def register_initial(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the beginning of a new transformer stage.
+
+        Args:
+            circuit: Input circuit to the new transformer stage.
+            transformer_name: Name of the new transformer stage.
+        """
+
+    @abc.abstractmethod
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+        """
+
+    @abc.abstractmethod
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+        """
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers."""
+
+    logger: Optional[TransformerStatsLoggerBase] = None
```

Comment:
Would we want to allow this to be optional ? Wondering if it would be better for things implementing this to always log at different levels and then leave it up to the user to set the level of reporting when they call a particular transformer.

Reply:
Since the logger is stateful, a user code which needs to use the logging when calling multiple optimization stages would need to explicitly pass a user generated `context` to each stage. i.e.

```python
>>> context = TransformerContext(logger)
>>> circuit = transformer_stage1(circuit, context)
>>> circuit = transformer_stage2(circuit, context)
...
```
whereas if they are not interested in the logs, they could simply do 
```python
>>> circuit = transformer_stage1(circuit)
>>> circuit = transformer_stage2(circuit)
```

assuming each transformer stage has a default `context=TransformerContext()` specified. If we make the logger as a required field, we'd expect users to always write way-1. I'm not sure if we should necessarily impose this restriction on users. LMK if you feel otherwise.

Reply:
Ahhh good point. Let's keep it optional.